rfcs/proposed/RFC-2-server-sdk-ncco-builders.md

```diff
@@ -0,0 +1,166 @@
+
+We can offer a comprehensive NCCO builder in each of the mainline Vonage SDKs that will ease the NCCO action composition burden on our users considerably. These builders will assist users in both creating their NCCO actions, and provide real-time feedback on incorrect construction. 
+
+## Proposed Solution
+
+### Technical Specifications
+
+The NCCO builder MUST be instantiable. By providing distinct instantiations for each NCCO action, it becomes easier to provide specific feedback on each unique action for the developer. For example, there may be multiple `type` parameters throughout the NCCO actions, but the specifications for `type` will be different for each unique NCCO action. The disaggregation and unique instantiation of each action as its own class allows for that fine attention to detail.
+
+```ruby
+# Standalone NCCO with two actions
+
+talk = Vonage::Voice::Ncco.talk(text: 'Hello World!')
+input = Vonage::Voice::Ncco.input(type: ['dtmf'], dtmf: { bargeIn: true })
+ncco = Vonage::Voice::Ncco.build(talk, input)
+
+# => [{:action=>"talk", :text=>"Hello World!"}, {:action=>"input", :type=>["dtmf"], :dtmf=>{:bargeIn=>true}}]
+```
+
+Each NCCO action MUST be able to be accessed via its own method name accessed on the NCCO class. For example, the `connect` action would be accessed as `Vonage::Voice::Ncco.connect` and the `notify` action would be accessed as `Vonage::Voice::Ncco.notify`. For example, the following would accomplish this in the Ruby SDK:
+
+```ruby
+	module Vonage
+  class Voice::Ncco
+    ACTIONS = {
+      connect: Vonage::Voice::Actions::Connect,
+      conversation: Vonage::Voice::Actions::Conversation,
+      input: Vonage::Voice::Actions::Input,
+      notify: Vonage::Voice::Actions::Notify,
+      record: Vonage::Voice::Actions::Record,
+      stream: Vonage::Voice::Actions::Stream,
+      talk: Vonage::Voice::Actions::Talk
+    }
+
+    ACTIONS.keys.each do |method|      
+      self.class.send :define_method, method do |attributes|
+        ACTIONS[method].new(**attributes).action
+      end
+    end
+
+    def self.method_missing(method)
+      raise ClientError.new("NCCO action must be one of the valid options. Please refer to https://developer.nexmo.com/voice/voice-api/ncco-reference#ncco-actions for a complete list.")
+    end
+
+    def self.create(*actions)
+      actions.flatten!
+    end
+  end
+end
+```
+
+The NCCO builder MUST NOT be part of the SDK client instantiation. Adding the NCCO builder to the client could cause confusion for users as the expectation for client methods is a direct interaction with the Vonage APIs. In this regard, it SHOULD be similar to the JWT generator as specified in its [RFC](https://raw.githubusercontent.com/Nexmo/server-sdk-specification/main/rfcs/accepted/RFC-1-server-sdk-jwt-packages.md).
+
+The NCCO builder MUST return a fully constructed NCCO in the form of an array, with each action represented as a hash inside the array.
+
+### Field Specifications
```

Comment:
I would remove the "default" value. I found when building the PHP NCCO's that passing the default value can in many cases be different behavior that not sending any value. I would change the heading to "Values" if they act as enums or have ranges and list the values (like Record.format)

Reply:
That's a really good point. I'll get that pushed up today.

Reply:
@dragonmantank, done ✅